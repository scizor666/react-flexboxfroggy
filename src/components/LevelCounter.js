import React from 'react';
import {connect} from 'react-redux';
import {changeLevel} from '../actions'


const LevelCounter = props => {
    const isFirstLevel = props.current === 1;
    const isLastLevel = props.current === props.max;

    const handleLevelDecrement = () => props.changeLevel(isFirstLevel ? 1 : props.current - 1);
    const handleLevelIncrement = () => props.changeLevel(isLastLevel ? props.max : props.current + 1);

    return <div className='levelCounter'>
        <span className={`levelCounter__left${isFirstLevel ? ' u-disabled' : ''}`} onClick={handleLevelDecrement}>◀</span>
        <span className='levelCounter__label'>Level {props.current} of {props.max} ▾</span>
        <span className={`levelCounter__right${isLastLevel ? ' u-disabled' : ''}`} onClick={handleLevelIncrement}>▶</span>
    </div>;
};

const mapStateToProps = ({level}) => ({...level});

export default connect(mapStateToProps, {changeLevel})(LevelCounter);
