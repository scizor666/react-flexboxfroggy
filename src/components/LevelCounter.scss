@mixin arrow {
  display: inline-block;
  padding: 0 0.4em;
  line-height: 30px;
  cursor: pointer;

  &:not(.u-disabled):hover {
      background-color: rgba(255, 255, 255, 0.4);
  }
}

$background-color: rgba(255, 255, 255, 0.2);

.levelCounter {
  align-self: center;
  border-radius: 4px;
  user-select: none;

  &__left {
    @include arrow;
    border-radius: 4px 0 0 4px;
    background-color: $background-color;
  }

  &__label {
    line-height: 30px;
    text-align: center;
    cursor: pointer;
    display: inline-block;
    width: 160px;
    border-left: 1px solid $background-color;
    border-right: 1px solid $background-color;
    background-color: $background-color;
    &:hover {
      background-color: rgba(255, 255, 255, 0.4);
    }
  }

  &__right {
    @include arrow;
    border-radius: 0 4px 4px 0;
    background-color: $background-color;
  }
}
