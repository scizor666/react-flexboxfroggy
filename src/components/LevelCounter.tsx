import React from 'react';
import {RouteComponentProps, withRouter} from 'react-router-dom';
import LevelContext from '../context/LevelContext';
import ILevel from "../types/ILevel";

const LevelCounter = (props: RouteComponentProps) => {

    return <LevelContext.Consumer>{(level: ILevel) => {
        const currentLevel = level.current === 'win' ? level.max : level.current; // you can make it better
        const isFirstLevel = currentLevel === 1;
        const isLastLevel = currentLevel === level.max;

        const handleLevelDecrement = () => props.history.push(`/level/${isFirstLevel ? 1 : currentLevel - 1}`);
        const handleLevelIncrement = () => props.history.push(`/level/${isLastLevel ? level.max : currentLevel + 1}`);

        return <div className='levelCounter'>
        <span className={`levelCounter__left${isFirstLevel ? ' u-disabled' : ''}`}
              onClick={handleLevelDecrement}>◀</span>
            <span className='levelCounter__label'>Level {currentLevel} of {level.max} ▾</span>
            <span className={`levelCounter__right${isLastLevel ? ' u-disabled' : ''}`}
                  onClick={handleLevelIncrement}>▶</span>
        </div>
    }}
    </LevelContext.Consumer>;
};

export default withRouter(LevelCounter);
