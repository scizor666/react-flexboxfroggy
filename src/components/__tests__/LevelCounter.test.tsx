import {mount} from 'enzyme';
import React from 'react';
import renderer from 'react-test-renderer';
import LevelCounter from "../LevelCounter";

const mockLevelContext = jest.fn();

jest.mock('../../context/LevelContext', () => ({
    Consumer: ({children}: any) => children(mockLevelContext())
}));

describe('Level Counter', () => {

    beforeEach(() => {
        mockLevelContext.mockReset();
    });

    [[1,24], [0, 90]].forEach(([from , to]) => {
        test(`renders level values as ${from} of ${to}`, () => {

            mockLevelContext.mockReturnValue({current: from, max: to});

            const wrapped = mount(React.createElement((LevelCounter as any).WrappedComponent)); // probably is a problem with @types for react-router
            expect(wrapped.find('.levelCounter__label').text()).toBe(`Level ${from} of ${to} â–¾`);

            const tree = renderer.create(React.createElement((LevelCounter as any).WrappedComponent));
            expect(tree.toJSON()).toMatchSnapshot();
        })
    })
});
