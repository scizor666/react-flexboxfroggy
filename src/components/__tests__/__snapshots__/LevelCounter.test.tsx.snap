// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Level Counter renders level values as 0 of 90 1`] = `
<div
  className="levelCounter"
>
  <span
    className="levelCounter__left"
    onClick={[Function]}
  >
    ◀
  </span>
  <span
    className="levelCounter__label"
  >
    Level 
    0
     of 
    90
     ▾
  </span>
  <span
    className="levelCounter__right"
    onClick={[Function]}
  >
    ▶
  </span>
</div>
`;

exports[`Level Counter renders level values as 1 of 24 1`] = `
<div
  className="levelCounter"
>
  <span
    className="levelCounter__left u-disabled"
    onClick={[Function]}
  >
    ◀
  </span>
  <span
    className="levelCounter__label"
  >
    Level 
    1
     of 
    24
     ▾
  </span>
  <span
    className="levelCounter__right"
    onClick={[Function]}
  >
    ▶
  </span>
</div>
`;
